CASSGO-50 Endless query execution fix proposal

tengu-alt opened the pull request:
I've researched possible endless query execution when `HostSelectionPolicy` returns the same downed host ([CASSGO-50](https://issues.apache.org/jira/browse/CASSGO-50)) and propose three approaches to prevent it.

- Wrap `hostIter` to prevent it from returning downed host before executing the query. 
 e.g.
```
  hostIter = func() SelectedHost {
    checkIter := hostIter()
    if !checkIter.Info().IsUp() {
      return nil
    }
    return checkIter
  }
```
But that approach can cause breaking changes for applications which already relies on user-defined policies that can return downed hosts.

- Properly documented `type NextHost func()`:
Add a comment for `NextHost` type describing that `NextHost` should return not downedhost or return nil

- Add the threshold limit based on the number of hosts inside `(q *queryExecutor) do` which will prevent the endless execution without the breaking changes

Implementations of the threshold and documentation are presented in this PR.

tengu-alt commented:
@joao-r-reis @jameshartig can you please take a look at this?

joao-r-reis commented:
Hmm I don't like the idea of introducing a dependency on the number of connection pools but wrapping `hostIter` won't solve the issue because whether the host is down or nil doesn't matter much, it will lead to the same behavior...

I actually think it's fine for the driver to assume that the host selection policy is behaving correctly (i.e. not returning the same host over and over) but we still have the issue that in #1793 we added an internal policy that does this (returns the same host everytime the function is called).

I've commented this before but I truly think the best option would be to make the internal host selection policy added in #1793 return the host only once even if that comes at a cost of having no retries or speculative executions when the user uses this feature.

Later down the line if we see that there is demand for retries/speculative executions when the featured added in #1793 is used then we can reevaluate.

One potential alternative if we absolutely want to keep retries when a specific host is being targeted by the query is to add a special flag so that the query executor treats this host selection policy differently (i.e. break from the loop immediately when the returned host is nil or downed).

joao-r-reis commented:
@tengu-alt do you have time to address [this](https://github.com/apache/cassandra-gocql-driver/pull/1875#issuecomment-2818077493)? We're trying to resolve the pending PRs for the 2.0 release

tengu-alt commented:
> @tengu-alt do you have time to address https://github.com/apache/cassandra-gocql-driver/pull/1875#issuecomment-2818077493? We're trying to resolve the pending PRs for the 2.0 release

Sorry for the late response. I've pushed another approach according to your suggestions.  
I propose the deletion of the pool after we got it using the `getPoolByHostID`, and keep the documentation about policies correct behavior.

joao-r-reis commented on connectionpool.go:
```diff
@@ -246,6 +246,9 @@ func (p *policyConnPool) getPool(host *HostInfo) (pool *hostConnPool, ok bool) {
 	p.mu.RLock()
 	pool, ok = p.hostConnPools[hostID]
+	if ok {
+		p.hostConnPools[hostID] = nil
```
What's the goal here? In my mind this will make it so that only one query will ever run on a specific pool no?

tengu-alt replied:
Good catch, fixed. I've changed the `hostIter` func so it returns the `SelectedHost` once.
.

joao-r-reis commented on query_executor.go:
```diff
@@ -89,14 +89,18 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
-			// we return nil to avoid endless query execution in queryExecutor.do()
-			if !ok || !pool.host.IsUp() {
-				return nil
+			if host != nil {
```
To ensure this `hostIter` can be safely called concurrently I think we should make the outer variable an `int` and then use `atomic.CompareAndSwap` in this `if`

```
		returnedHostOnce := 0
		hostIter = func() SelectedHost {
			if atomic.CompareAndSwapInt32(&returnedHostOnce, 0, 1) {
                            return return (*selectedHost)(host)
                        }
                        return nil
                }
```

tengu-alt replied:
Understood, refactored.

joao-r-reis commented on query_executor.go:
```diff
@@ -89,14 +89,18 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 	// if it is, the query should be executed at the corresponding host.
 	if hostID := qry.GetHostID(); hostID != "" {
+		pool, ok := q.pool.getPoolByHostID(hostID)
+		if !ok || !pool.host.IsUp() {
```
I don't think we need this now that we return `nil` after the first execution since this check is already being done in the query executor. We needed this before because we needed a way to prevent the `hostIter` from returning the same unresponsive host over and over again.

tengu-alt replied:
Got it, changed.

joao-r-reis commented on policies.go:
```diff
@@ -323,6 +323,7 @@ func (host *selectedHost) Info() *HostInfo {
 func (host *selectedHost) Mark(err error) {}
 
 // NextHost is an iteration function over picked hosts
+// Should return nil if SelectedHost is not Up to prevent endless query execution.
```
the driver checks if the host is not up, the function should "eventually" return nil regardless of what causes it.

`// Should return nil eventually to prevent endless query execution.`

tengu-alt replied:
Changed.

joao-r-reis commented on query_executor.go:
```diff
@@ -89,14 +90,17 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
+		pool, ok := q.pool.getPoolByHostID(hostID)
+		if !ok {
+			return nil, ErrNoConnections
+		}
```
remove lines 93-96, it's not needed as I explained before

tengu-alt replied:
I don't clearly understand: we need the pool to get the host from it.

joao-r-reis replied:
oh sorry hmm I think it would make more sense to just get the host from the ring here, I didn't even realize it was only getting the pool to get the host before.
Can you replace this with a call to `session.ring.getHost(id)`? Maybe change `getHost` to return `host, bool` so that we can check whether the lookup failed here and return an error

tengu-alt replied:
Got it, refactored

joao-r-reis commented on query_executor.go:
```diff
@@ -89,14 +90,17 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
-			if !ok || !pool.host.IsUp() {
-				return nil
+			if atomic.CompareAndSwapInt32(&returnedHostOnce, 0, 1) {
+				return (*selectedHost)(host)
```
we should add a test to ensure that if a host id is set on the query then it only gets tried once even if it fails

tengu-alt replied:
As I checked, `TestQuery_SetHostID` in the `cassandra_test.go` already covers this case.

joao-r-reis reviewed (approved):
👍 nice, can you add to the changelog and format the commit message? Then I'll merge

tengu-alt committed:
Endless query execution fix

Fix for the endless query execution when HostSelectionPolicy returns
the same downed host.
Internal HostSelectionPolicy will return the host only once if HostID is set.
Documentation for external policies was added.

patch by Oleksandr Luzhniy; reviewed by João Reis, for CASSGO-50

tengu-alt commented:
> 👍 nice, can you add to the changelog and format the commit message? Then I'll merge

Done.
